### ISSUE TITLE
"Invalid toolchain" error with WSL2

### ISSUE THREAD

**Issue by Agrailag** [intellij-rust/intellij-rust#7311]
## Environment

* **IntelliJ Rust plugin version:** 0.4.147.3871-211
* **Rust toolchain version:** 1.54.0-nightly (1c6868aa2 2021-05-27) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2021.1.2
* **Operating system:** Windows 10 10.0

## Problem description

I select cargo location, it show correct version, but also it show error about invalid toolchain location. So i can't create new project.
![Intellij](https://user-images.githubusercontent.com/37513902/120847552-52c21600-c57c-11eb-9093-913eac2df6b2.png)

**Comment by Agrailag** [comment]
Ok, i find out that i forgot to select experimental feature, but it would be nice to have more helpful error for this one.

**Comment by Undin** [comment]
Thanks for the report!
The issue will be automatically resolved when we turn on the corresponding feature by default.
Don't think we should add additional user-friendly messages for experimental features

**Comment by Zokol** [comment]
I ran into this issue after updating rustup and cargo on WSL2.

```
$ rustc -V && cargo -V
rustc 1.52.1 (9bc8c42bb 2021-05-09)
cargo 1.52.0 (69767412a 2021-04-21)
```

![image](https://user-images.githubusercontent.com/1899518/122044387-1c3e9380-cde5-11eb-90da-a9cc7b01c8cc.png)

**Comment by Undin** [comment]
@Zokol Is `org.rust.wsl` [experimental feature](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features) enabled?

**Comment by Zokol** [comment]
> @Zokol Is `org.rust.wsl` [experimental feature](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features) enabled?

Thanks, this resolved the issue!

**Comment by jwilson8767** [comment]
I am also getting this message in IntelliJ IDEA, and I have enabled `org.rust.wsl`.


IntelliJ IDEA 2021.1.2 (Ultimate Edition)
Build #IU-211.7442.40, built on May 31, 2021
Runtime version: 11.0.11+9-b1341.57 amd64
VM: Dynamic Code Evolution 64-Bit Server VM by JetBrains s.r.o.
Windows 10 10.0
GC: ParNew, ConcurrentMarkSweep
Memory: 4933M
Cores: 8
Registry: terminal.buffer.max.lines.count=50000
Non-Bundled Plugins: be.jbeckers.compare_tab_with_editor2 (1.0.7), org.antlr.intellij.plugin (1.16), org.intellij.plugins.hcl (0.7.10), org.toml.lang (0.2.148.3911-211), String Manipulation (8.15.203.000.3), com.intellij.plugin.adernov.powershell (2.0.7), com.dmarcotte.handlebars (211.6693.44), izhangzhihao.rainbow.brackets (6.18), Pythonid (211.7442.45), ru.meanmail.plugin.requirements (2021.4-211), org.rust.lang (0.4.148.3911-211)
Kotlin: 211-1.4.32-release-IJ7442.2

![image](https://user-images.githubusercontent.com/10873576/122423357-dbb55600-cf5b-11eb-8a68-6fddcd0c81b9.png)


![image](https://user-images.githubusercontent.com/10873576/122423451-ed96f900-cf5b-11eb-8353-0bb7cbcf4ae3.png)

**Comment by mchernyavsky** [comment]
@jwilson8767 Does your Windows supports `\\wsl$\` paths? It requires version 1903.

**Comment by jwilson8767** [comment]
It does, I just confirmed again that this issue is still present. Here are the versions of all the relevant software I am currently running:
```
IntelliJ IDEA 2021.2 (Ultimate Edition)
Build #IU-212.4746.92, built on July 27, 2021
Runtime version: 11.0.11+9-b1504.13 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 10 10.0 (version: 20H2 build: 19042.1165)
GC: ParNew, ConcurrentMarkSweep
Memory: 4933M
Cores: 8
Non-Bundled Plugins: be.jbeckers.compare_tab_with_editor2 (1.0.7), org.antlr.intellij.plugin (1.16), org.intellij.plugins.hcl (0.7.10), org.toml.lang (0.2.154.4050-212-nightly), String Manipulation (8.15.203.000.3), com.intellij.plugin.adernov.powershell (2.0.7), com.dmarcotte.handlebars (212.4746.57), Pythonid (212.4746.96), ru.meanmail.plugin.requirements (2021.4.1-212), izhangzhihao.rainbow.brackets (6.19), org.rust.lang (0.4.154.4057-212-nightly)
Kotlin: 212-1.5.10-release-IJ4746.92


WSL instance:
Instance Name: ubuntu1804
OS: Ubuntu (18.04.5 LTS)
cargo path: /home/jwilson/.cargo/
```

I did manage to get it to let me hit OK if I enter my rust standard lib path as `\\wsl$\ubuntu1804\home\jwilson\.rustup\toolchains\stable-x86_64-unknown-linux-gnu\lib\rustlib` ...but it doesn't appear to save as re-opening the settings window shows it empty again.

**Comment by jaysonsantos** [comment]
Hey there, the first part seems to be fixed by this:
```diff
Index: src/main/kotlin/org/rust/cargo/util/ToolchainUtil.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/kotlin/org/rust/cargo/util/ToolchainUtil.kt b/src/main/kotlin/org/rust/cargo/util/ToolchainUtil.kt
--- a/src/main/kotlin/org/rust/cargo/util/ToolchainUtil.kt	(revision 6e5037ede001a61e3460203a45cc7b6b3ee716f4)
+++ b/src/main/kotlin/org/rust/cargo/util/ToolchainUtil.kt	(date 1639045951829)
@@ -5,13 +5,15 @@
 
 package org.rust.cargo.util
 
+import com.intellij.execution.wsl.WslDistributionManager
 import com.intellij.openapi.util.SystemInfo
 import org.rust.stdext.isExecutable
 import java.nio.file.Path
 
+
 fun Path.hasExecutable(toolName: String): Boolean = pathToExecutable(toolName).isExecutable()
 
 fun Path.pathToExecutable(toolName: String): Path {
-    val exeName = if (SystemInfo.isWindows) "$toolName.exe" else toolName
+    val exeName = if (SystemInfo.isWindows && !WslDistributionManager.isWslPath(this.toString())) "$toolName.exe" else toolName
     return resolve(exeName).toAbsolutePath()
 }
```
but then on `org.rust.cargo.toolchain.wsl.UtilsKt#hasExecutableOnWsl` in my case, it calls `pathToExecutableOnWsl(toolName).toFile().isFile` where `toolName` is `\\wsl$\Arch\usr\bin\rustc` and on arch it is a link, as you can see here:
```
PS C:\Users\santo\src\intellij-rust> dir \\wsl$\Arch\usr\bin\rustc


    Directory: \\wsl$\Arch\usr\bin


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-----l         11/9/2021   3:40 PM             15 rustc
```
I tried to fiddle with that function a bit like checking if the file exists, and it is not a directory, but for jvm that file is not there at all

**Comment by benWindsorCode** [comment]
I too am facing this issue. org.rust.wsl enabled, rust installed via rustup, toolchain location pointed to '\\wsl$\Ubuntu-20.04\home\ben\.cargo\bin' detects 'Toolchain version: 1.58.1' but then doesn't detect standard library and says 'invalid toolchain location: can't find Cargo in \home\ben\.cargo\bin'

windows build 19044
rustc 1.58.1
cargo 1.58.0
rustup 1.24.3
intellij 2021.3.1

**Comment by mxk** [comment]
Has anyone managed to get this working? Getting the same problem on CLion 2022.2.1, Rust plugin 0.4.176.4815-222. `org.rust.wsl` is enabled and it does not seem to be caused by rustc being a link.

**Comment by benWindsorCode** [comment]
yes +1 to this, has been outstanding for a while. I actually brought the all products pack because wanted to code with the rust wsl integration and its somewhat broken it seems

**Comment by w-biggs** [comment]
Having the exact same issue. Can't use CLion for Rust work because of this - cannot tell what's causing it at all. The experimental feature is enabled, putting the path to the standard library in manually does not fix the issue, and my rustc does not seem to be a link. I'm on plugin version 0.4.179.4903-222.

![image](https://user-images.githubusercontent.com/3406010/194716989-17740a1b-51e0-42b2-87af-18d019adc80a.png)

**Comment by benWindsorCode** [comment]
Still having this issue in Clion 2022.1.2 build #CL-221.5787.29, rust plugin 0.4.179.4903-221, rust version 1.65
